fix(requirements): refactor to use common codepath for table coalescing

ferantivero opened the pull request:
closes: #7045

### what?
regression has been introduced from dda84976bf75e4f3fca9090f0aabe6a7208d709d around the [exports format](https://v2.helm.sh/docs/developing_charts/#using-the-exports-format) behavior, resulting in child values not being propagated to parent's chart values.

### this PR

  - align both formats behaviors and now they will just differ in how to discover their paths
  - add coverage for exports format and fix expected assertions for [parent-child format](https://v2.helm.sh/docs/developing_charts/#using-the-child-parent-format) to match the logic child values always wins
  - just partially revert dda84976bf75e4f3fca9090f0aabe6a7208d709d, this way parents values could be overridden when coalescing
  - after getting better coverage we were able to refact both formats behaviors by merging their propagation logics into a single code path.

bacongobbler commented:
Hi there! A few things here need to be squared away before we can review.

1. Would you mind writing a test to cover #7045?
1. You need to sign off on the DCO before we can accept the PR. More info in the failed check: https://github.com/helm/helm/pull/7047/checks?check_run_id=314807683
1. A longer explanation about the change and why it has to be reverted might be nice. I don't like to revert changes if at all possible, as it will reintroduce the same issue #5821 was trying to solve. Have you considered alternative approaches?

Commit:
partially revert refactor for table coalescing

go back the empty initialization of map values, otherwise parent chart
values wins when calescing.

Signed-off-by: ferantivero <v-fean@microsoft.com>

Commit:
align export format behavior with parent-child format

now export format will coalesce tables in the very same way as in
parent-child format.

  - now both formats follow the leaft chart (subchart) wins. It means,
  values in child charts will propagate their values to parents charts

solved: #7045
Signed-off-by: ferantivero <v-fean@microsoft.com>

Commit:
fix unit tests for parent-child format

  - change expected values to be aligned with child propagation approach.
    Leaft charts always win. It means, it overrides the parent's chart
    values always
  - improve expected/actual output unit test info to easily understand where the
    assertion fails

Signed-off-by: ferantivero <v-fean@microsoft.com>

Commit:
add unit test for export format

   - add coverage for cannonical export format
   - add new subchart3 for testing purposes

solved: #7045
Signed-off-by: ferantivero <v-fean@microsoft.com>

ferantivero commented:
thanks for your feedback @bacongobbler 

> Would you mind writing a test to cover #7045?

done 

You need to sign off on the DCO before we can accept the PR. 

done

> A longer explanation about the change and why it has to be reverted might be nice. 

When a dependency (requirements.yaml) for a chart attempts to propagate its values (data) to override parent's chart values, it results in being ignored.

Please let's take a look at the commit messages for further details and also please let's have a look at the #7045 where more information, repro steps, and chart details were provided.

> I don't like to revert changes if at all possible, as it will reintroduce the same issue #5821 was trying to solve. 

we got a partial revert now 😸

> Have you considered alternative approaches?

Yes, so now we aligned both formats behavior with the approach leaf-sub-chart-values-always-win. Therefore, sub chart values override the parent values, what I think is now pretty much aligned with [docs](https://v2.helm.sh/docs/developing_charts/#importing-child-values-via-requirements-yaml) and how `exports format` has been working till v2.15.0.

Commit:
refact to materialize the propagation logic

  - formats exports and parent-child just differ in how to discover
    their paths
  - share propagation logic for both formats

Signed-off-by: ferantivero <v-fean@microsoft.com>

Commit:
remove dup expected entries

Signed-off-by: ferantivero <v-fean@microsoft.com>

Commit:
remove empty lines

Signed-off-by: ferantivero <v-fean@microsoft.com>

ferantivero commented:
ping @bacongobbler, friendly reminder 😃

ferantivero commented:
@bacongobbler just been testing some of our charts using helm3, and I wanted to confirm that this PR is fully aligned with Helm 3 behavior. In other words, #7045 is not present in Helm 3.

technosophos reviewed (approved):
I'm going to beg @adamreese to double check (since he did the work on this portion in Helm 3), but this looks good to me.

I am provisionally putting this on 2.16.4.

adamreese reviewed (approved):
Great refactor!
